Mesocanonical ensemble Monte Carlo

shivamrkparashar opened the pull request:
**Summary**

This feature adds the mesocanonical ensemble Monte Carlo (MCEMC) simulations to calculate adsorption isotherms. The MCEMC complements the grand canonical Monte Carlo (GCMC) but provides more details like the location of vapor-side and liquid-side spinodals, and vapor-liquid equilibrium. MCEMC produces a canonical type S-shaped isotherm in contrast to hysteretic GCMC isotherm for mesoporous materials.

**Related Issue(s)**

No known issues at this time

**Author(s)**

Shivam Parashar (Rutgers University)
Email: shivamrkparashar@gmail.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Artificial Intelligence (AI) Tools Usage**

By submitting this pull request, I confirm that I did NOT use any AI tools to generate
all or parts of the code and modifications in this pull request.

**Backward Compatibility**

No known issues with backward compatibility

**Implementation Notes**

MCEMC simulations are similar to GCMC so the MCEMC is implemented as an option in fix gcmc. MCEMC requires two boxes (system and gauge cell), but only one (the system) is actually simulated. The gauge cell is assumed to be an ideal gas and need not be simulated. An example to run a simple LJ fluid is added, similar to GCMC example.

The correctness of MCEMC simulations is verified by comparing against GCMC simulations in this repository- https://github.com/shivamrkparashar/lammps-simulations 

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**
Link to the MCEMC paper- https://doi.org/10.1016/j.jcis.2024.06.083

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+* ID, group-ID are documented in :doc:`fix <fix>` command
+* mcemc = style name of this fix command
+* N = invoke this fix every N steps
+* X = average number of GCMC exchanges to attempt every N steps
```
Shouldn't this say MCEMC?

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+* type = atom type (1-Ntypes or type label) for inserted atoms (must be 0 if mol keyword used)
+* seed = random # seed (positive integer)
+* T = temperature of the ideal gas reservoir (temperature units)
+* vgauge = gauge volume for GCMC exchanges (volume units)
```
same as above

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+       *shake* value = fix-ID
+         fix-ID = ID of :doc:`fix shake <fix_shake>` command
+       *region* value = region-ID
+         region-ID = ID of region where GCMC exchanges and MC moves are allowed
```
more

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+       *region* value = region-ID
+         region-ID = ID of region where GCMC exchanges and MC moves are allowed
+       *maxangle* value = maximum molecular rotation angle (degrees)
+       *full_energy* = compute the entire system energy when performing GCMC exchanges and MC moves
```
one more

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+simulation by exchanging particles with a finite volume ideal gas reservoir (gauge cell) at
+the same temperature as the system as discussed in :ref:`(Parashar) <Parashar>`.
+It also attempts Monte Carlo moves (translations and rotations) of particles
+within the simulation cell. Specific use of this fix is to compute adsorption
```
```suggestion
within the simulation cell. Specific uses of this fix are to compute the adsorption
```

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+the same temperature as the system as discussed in :ref:`(Parashar) <Parashar>`.
+It also attempts Monte Carlo moves (translations and rotations) of particles
+within the simulation cell. Specific use of this fix is to compute adsorption
+isotherm in porous materials, or computing vapor-liquid equilibrium of fluids.
```
```suggestion
isotherm in porous materials, or to compute the vapor-liquid equilibrium of fluids.
```

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+isotherms for microporous materials. But for large pores (> 2 nm), GCMC gives
+a hysteretic adsorption/desorption isotherm, while MCEMC gives a reversible S-
+shaped van der Waals type isotherm, as discussed in :ref:`(Parashar) <Parashar>`.
+MCEMC isotherm spans the stable and metastable states, while GCMC samples only the
```
```suggestion
The MCEMC isotherm spans the stable and metastable states, while GCMC samples only the
```

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+a hysteretic adsorption/desorption isotherm, while MCEMC gives a reversible S-
+shaped van der Waals type isotherm, as discussed in :ref:`(Parashar) <Parashar>`.
+MCEMC isotherm spans the stable and metastable states, while GCMC samples only the
+stable states. The MCEMC is a middle ground between the grand canonical ensemble
```
```suggestion
stable states. The MCEMC method is a middle ground between the grand canonical ensemble
```

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+MCEMC isotherm spans the stable and metastable states, while GCMC samples only the
+stable states. The MCEMC is a middle ground between the grand canonical ensemble
+which permits unlimited fluctuations, and the canonical ensemble, which
+considers a close system. The MCEMC simulations generate the adsorption
```
```suggestion
considers a closed system. MCEMC simulations generate adsorption
```

akohlmey commented on doc/src/fix_mcemc.rst:
```diff
@@ -0,0 +1,509 @@
+The *full_energy* option means that the fix calculates the total
+potential energy of the entire simulated system, instead of just
+the energy of the part that is changed. The total system
+energy before and after the proposed GCMC exchange or MC move
```
```suggestion
energy before and after the proposed MCEMC exchange or MC move
```

akohlmey reviewed (changes requested):
I have only looked at the documentation and there are some small (grammar) improvements suggested and a few clarifications mentioned.

akohlmey commented on examples/mc/in.mcemc.lj:
```diff
@@ -0,0 +1,59 @@
+
+# run
+
+run             10000
```
FYI: this file is missing a newline character on the final line. This can lead to the line not being read.

shivamrkparashar commented:
I have addressed the issues in the latest commits. Thanks for your feedback Axel. Please let me know if I missed something.

akohlmey commented:
> I have addressed the issues in the latest commits. Thanks for your feedback Axel. Please let me know if I missed something.

There are four more additional formal changes required to complete the full integration:
- the documentation source should have a line `.. versionadded:: TBD` followed by an empty line right before the paragraph titled `Description`
- add the names of the two source files to `src/.gitignore` (like for the other files in the `src/MC` folder)
- add a reference to the doc file to `doc/src/Commands_fix.rst`
- add a reference and a one-line description to the doc file `doc/src/fix.rst`

Please note that somebody else will have to review the source code and the science etc.

shivamrkparashar commented:
Thanks again Axel for your valuable suggestions. I have pushed a commit that resolves them.

sjplimp commented:
Thanks for contributing this pull request with a new feature.  A few comments
(a) I'd like to have Aidan review this PR as he is the author of fix gcmc, and can advise on (b,c)
(b) If the new source files duplicate a lot of code in fix gcmc (from the doc page it appears the syntax is nearly identical),
I wonder if this would be better implemented as a derived class of fix gcmc, with possibly only a few of the methods needing to be overwritten.  Or possibly as a new option for the fix gcmc command.
(c) The explanation above mention using 2 boxes, but I am not sure what that means.  And I don't see anything about that in the command doc page.  Possibly you only mean 2 boxes in a conceptual sense?  Please explain further.

sjplimp commented:
@athomps My previous comment should have included you

akohlmey commented on src/MC/fix_mcemc.h:
```diff
@@ -0,0 +1,158 @@
+  double reservoir_temperature;
+  double tfac_insert;
+  double gaugecell_volume;
+  double ntotal;
```
What is ntotal representing? How is it different from `Atom::natoms`? and why is it a `double` and not an `int` like `ngas` or a `bigint` like `Atom::natoms`?

shivamrkparashar replied:
Thanks, Axel, for pointing this out. I have updated ntotal to be an int in the latest commits.

shivamrkparashar commented:
Thanks, Steve, for your valuable suggestions. I am now deriving mcemc from gcmc, and it avoids some code duplication. 
MCEMC uses two boxes: the simulation cell and the gauge cell. Only the former is explicitly simulated in LAMMPS, whereas the latter is assumed to be an ideal gas. To make this clear, I have also updated the description of this pull request.

sjplimp reviewed (changes requested):
see my latest comment

akohlmey commented:
> To make this clear, I have also updated the description of this pull request.

@shivamrkparashar Please note that this is insufficient. It should be explained in the documentation of the fix command. The pull request description will be invisible to users that just download LAMMPS or look at the online documentation.

sjplimp commented:
Thanks @shivamrkparashar for reformulating the new command to derive from the existing fix gcmc.  I have just a couple more hi-level comments about the new capability.  I also still think Aidan should review this PR before merging.

(1) For a new derived class, the normal LAMMPS naming convention would be something like fix gcmc/meso and fix_gcmc_meso.cpp/h.  But in this case, if MCEMC is a more well-known descriptor, it would be OK to leave it as is I think.

(2) Similar to the previous version of the code, I think the new doc page fix_mcemc.rst, duplicates a lot of info on the fix gcmc.rst doc page ?  There are 2 common ways to avoid this in LAMMPS doc pages.

(a) Both commands share a single doc page, which lists both commands at the top (see the fix rigid doc pages).  There could be a new section of this modified doc page which explains MCEMC and the difference vs GCMC.  Ditto for any new fix mcemc options which are unique to MCEMC or are used differently than how they are used for GCMC.

(b) Have a more streamlined fix MCEMC doc page which states that all (or most) of the fix gcmc options can be used and simply refers to that doc page for their syntax and explanation.  Only new options unique to MCEMC (or changed from their GCMC usage) are listed on its doc page.  See the write_dump doc page for an example of this.  It does not reproduce all the syntax of the dump and dump_modify commands.

sjplimp commented:
The goal of my comment (2) is to reduce duplication in the doc pages and make it easier for users to read and find out what they need to know about MCEMC.

shivamrkparashar commented:
Thanks @akohlmey and @sjplimp for your valuable comments.

@akohlmey, I have mentioned in the documentation that the system is explicitly simulated, but the gauge cell is not. 

@sjplimp to answer your comments-
1. Yes, MCEMC is derived from GCMC because of its similarities and to avoid code duplication. From a user's perspective, it is a method parallel (or complementary) to GCMC. Therefore, I think keeping the name `fix mcemc` would be suitable.
2. I agree. The fix_mcemc documentation was copied and modified from the fix_gcmc documentation, and it was challenging for the users to identify what is new in mcemc from its current state. Therefore, I have removed the duplications from the MCEMC documentation and kept only the new details relevant to MCEMC. I hope this will make users' lives easier.

athomps reviewed (changes requested):
Thank you for this contribution. Please accept our apologies for the slow response.  This finite reservoir variant of GCMC is a nice addition and we would like to include it.  However, since most of the code is identical to the original, it would make more sense to make it available to users via an optional keyword on `fix gcmc`.  This would allow definition of the `Vgauge` and `Ngauge` values.  A flag could then be set in `fix_gcmc.cpp` than would conditionally use the modified MC move acceptance criterion.

shivamrkparashar commented:
Hi @athomps. I appreciate your suggestion. I want to keep fix mcemc separated, as it is one of the methods to calculate the adsorption isotherm alongside GCMC and the canonical ensemble Monte Carlo (or NVE). Unlike the grand canonical ensemble, which assumes an infinite reservoir and a chemical potential as input, the mesocanonical ensemble uses a finite-volume reservoir, which measures chemical potential during the simulation. Keeping it within fix gcmc (as an optional keyword) would blur the distinction and is more likely to be misunderstood and misused.

athomps commented:
> Hi @athomps. I appreciate your suggestion. I want to keep fix mcemc separated, as it is one of the methods to calculate the adsorption isotherm alongside GCMC and the canonical ensemble Monte Carlo (or NVE). Unlike the grand canonical ensemble, which assumes an infinite reservoir and a chemical potential as input, the mesocanonical ensemble uses a finite-volume reservoir, which measures chemical potential during the simulation. Keeping it within fix gcmc (as an optional keyword) would blur the distinction and is more likely to be misunderstood and misused.

The primary consideration here is code quality.  Any issues with misunderstanding and misuse is secondary, and can be mitigated by adding good explanations and examples to the documentation.

athomps commented:
Given that this PR needs to be substantially reworked, I have converted the status to draft.

athomps reviewed (changes requested):
This feature should be made into an option on the `fix gcmc`command, instead of a separate command. As discussed in the comments, the primary consideration is code quality. Any issues with misunderstanding and misuse is secondary, and can be mitigated by adding good explanations and examples to the documentation.

shivamrkparashar commented:
Hi @athomps, it may not be conceptually appropriate to include MCEMC as an option within fix gcmc, since the two operate in fundamentally different ensembles. In that respect, the situation is analogous to fix nvt and fix npt in LAMMPS, which are implemented as distinct commands rather than as options under a single fix. One possible improvement that I can see now is to keep a common doc page for fix gcmc and fix mcemc, similar to the fix nvt and fix npt pages (as recommended by @sjplimp).

akohlmey commented:
> Hi @athomps, it may not be conceptually appropriate to include MCEMC as an option within fix gcmc, since the two operate in fundamentally different ensembles. In that respect, the situation is analogous to fix nvt and fix npt in LAMMPS, which are implemented as distinct commands rather than as options under a single fix. One possible improvement that I can see now is to keep a common doc page for fix gcmc and fix mcemc, similar to the fix nvt and fix npt pages (as recommended by @sjplimp).

@shivamrkparashar Please note that @athomps is the principal author of `fix gcmc`. So when he requests this kind of change, there is no room left for arguing. Please also note that your example of `fix nvt` and `fix npt` is a surprisingly bad example, since those are actually a *single* implementation (in the `fix_nh.cpp` source file) and the individual commands are mainly retained for backward compatibility as minimal wrappers (by coincidence, it is also @athomps that refactored the thermostat and barostat fixes into the current layout). Finally a personal observation: it would actually be better to have a single `fix nh` command (or `fix nose-hoover`). The current nomenclature compels people to incorrectly claim to run in the NVE, NVT, NPT or similar ensemble (when they are, in fact, not) simply because they use one of the fixes with the corresponding name.

Bottom line: none of the LAMMPS developers will override the request for changes so if you *do* want your contribution to be included, you will have to make the requested changes. Otherwise the pull request will remain unmerged until we close it for old age. Sorry, if that is inconvenient for you. These are the kind of policies that we as a team of developers have agreed on.

shivamrkparashar committed:
Add mcemc as an option in fix gcmc

shivamrkparashar committed:
Fix whitespace

shivamrkparashar committed:
Add citation

shivamrkparashar committed:
Add example output files

shivamrkparashar committed:
Fix molecule ins/del probabilities

shivamrkparashar committed:
Improve documentation

shivamrkparashar committed:
Define ngauge

shivamrkparashar commented:
Hello @akohlmey and @athomps, in the recent changes I have implemented the mcemc method as an option to fix gcmc, as was suggested. I look forward to your valuable feedback. Thank you!

akohlmey commented on doc/src/fix_gcmc.rst:
```diff
@@ -69,23 +73,25 @@ Examples
 Description
 """""""""""
 
-This fix performs grand canonical Monte Carlo (GCMC) exchanges of atoms
-or molecules with an imaginary ideal gas reservoir at the specified *T*
-and chemical potential (:math:`\mu`) as discussed in :ref:`(Frenkel)
-<Frenkel2>`.  It also attempts Monte Carlo (MC) moves (translations and
+The fix can operate in two different ensembles: grand canonical ensemble
+Monte Carlo (GCMC) - particle exchange with infinite reservoir as discussed in :ref:`(Frenkel)
+<Frenkel2>` or mesocanonical ensemble Monte Carlo (MCEMC) - particle exchange
+with finite gauge cell as discussed in :ref:`(Parashar) <Parashar>`.
+The choice of ensemble is determined by the presence or absence of the *mcemc*
+keyword in the fix command. It also attempts Monte Carlo (MC) moves (translations and
 molecule rotations) within the simulation cell or region.  If used with
```
This should be `.. versionadded::` for an added feature. `.. versionchanged::` is used to document keywords that have their behavior changed.

shivamrkparashar replied:
Added this

akohlmey commented on examples/mc/log.15Mar26.mcemc.lj.g++.4:
```diff
@@ -0,0 +1,233 @@
+LAMMPS (11 Feb 2026 - Development - e4d534cd9b-modified)
```
Indeed if runs are updated with all pending changes committed, it allows to reproduce them by looking up the commit hash.

shivamrkparashar replied:
Fixed this

shivamrkparashar committed:
Minor fixes

sjplimp reviewed (approved):
@shivamrkparashar Thanks again for this contribution of a new MC algorithm.  And thanks for reformulating it several times in repsonse to reviewer comments.

@akohlmey I have approved the PR, but I think @athomps should make the final approval.

athomps commented on doc/src/fix_gcmc.rst:
```diff
@@ -25,7 +25,10 @@ Syntax
 
-     keyword = *mol* or *mcmoves* or *rigid* or *shake* or *region* or *maxangle* or *pressure* or *fugacity_coeff* or *full_energy* or *charge* or *group* or *grouptype* or *intra_energy* or *tfac_insert* or *overlap_cutoff* or *max* or *min*
+     keyword = *mcemc*, *mol*, *region*, *maxangle*, *pressure*, *fugacity_coeff*, *full_energy*, *charge*, *group*, *grouptype*, *intra_energy*, *tfac_insert*, or *overlap_cutoff*
+       *mcemc* values = ntotal vgauge
```
What happened to *mcmoves*, *rigid*, *shake*, *max*, and *min*?

athomps commented on src/MC/fix_gcmc.cpp:
This is very disruptive for the original code. The same effect could be obtained using something like:
if (run_mcemc) zz = exp((mccmc_ngauge+1)+log(mccmc_vguage))

athomps reviewed (changes requested):
@shivamrkparashar Thank you for attempting to address our previous requests that this feature be added as an option to `fix gcmc`.  Unfortunately, I think you are going to have to start over and try again.   From reading your changes to the documentation and `fix_gcmc.cpp`, you seem to have adopted the view that the GCMC and MCEMC should be presented to the users and implemented in the code as two equally important alternatives.  If you want to get your code in to LAMMPS, you are going to have to shift your perspective.  In contrast to the widely used GCMC method, MCEMC occupies a small niche in the molecular simulation ecosystem.  Therefore, the footprint in the documentation and the code also needs to be made as small as possible.  It is particularly important that you not break existing functionality.  The best way to avoid this is by limiting your changes to the existing code to things that are unavoidable.  Also, it is important not to disrupt the existing documention, which we have carefully built up over many years.  In making your changes, you seem to have deleted several optional keywords in the list, including *mcmoves*, *rigid*, *shake*, *max*, and *min*.  In the same spirit of minimal disruption. I ask that the introductory statement be left unchanged, but with a small qualifying statement.
```
This fix performs grand canonical Monte Carlo (GCMC) exchanges of atoms
or molecules with an imaginary ideal gas reservoir at the specified *T*
and chemical potential (:math:`\mu`) as discussed in :ref:`(Frenkel)
<Frenkel2>`.  There is also an option to perform mesocanonical ensemble Monte Carlo (MCEMC) - particle exchange with finite gauge cell as discussed in :ref:`(Parashar) <Parashar>`, see *mcemc* keyword below.
```

Similarly, the pervasive use of `if (run_mcemc) then ...else... ` construct is unacceptable.  Much less disruptive would be to minimize changes to existing code by using existing variables, as I indicated on line 942 (if (run_mcemc) zz = exp((mccmc_ngauge+1)+log(mccmc_vguage)))

Again, it is hard to overstate the importance of not breaking or muddying up things that are already working.  The goal is not to make things easier for you, nor for users of MCEMC, but to minimize disruption for developers and users of GCMC.

akohlmey commented:
@shivamrkparashar Looking over the various comments and discussions it seems to me that it would be better to close this pull request and start from scratch based on the recommendations from @athomps taking the current implementation as a reference and re-implementing in the more minimal and less invasive style requested. I am sorry for giving you the "runaround" but this sometimes happens, unfortunately, since not all of us are fully in tune with all parts of LAMMPS and @athomps *is* the authority on fix gcmc.
